vnext/ReactUWP/Modules/AsyncStorageModuleWin32.cpp

```diff
@@ -0,0 +1,191 @@
+      if (!key)
+        checkSQLiteResult(sqlite3_errcode(m_db));
+      auto value = reinterpret_cast<const char *>(sqlite3_column_text(pStmt.get(), 1));
+      if (!value)
+        checkSQLiteResult(sqlite3_errcode(m_db));
+      result.push_back(folly::dynamic::array(std::string(key), std::string(value)));
+      checkSQLiteResult(sqlite3_reset(pStmt.get()));
+    }
+  }
+  callback({folly::dynamic(), result});
+}
+
+void AsyncStorageModuleWin32::multiSet(folly::dynamic params, Callback callback) {
+  auto &&args = params[0];
+  if (args.size() != 0) {
+    Sqlite3Transaction transaction(m_db);
+    auto pStmt = PrepareStatement(u8"INSERT OR REPLACE INTO AsyncLocalStorage VALUES(?, ?)");
+    for (auto &&arg : args) {
+      BindString(pStmt, 1, arg[0].getString());
+      BindString(pStmt, 2, arg[1].getString());
+      checkSQLiteResult(sqlite3_step(pStmt.get()));
+      checkSQLiteResult(sqlite3_reset(pStmt.get()));
+    }
+  }
+  callback({});
+}
+
+void AsyncStorageModuleWin32::multiRemove(folly::dynamic params, Callback callback) {
+  auto &&args = params[0];
+  if (args.size() != 0) {
+    Sqlite3Transaction transaction(m_db);
+    auto pStmt = PrepareStatement(u8"DELETE FROM AsyncLocalStorage WHERE key = ?");
+    for (auto &&arg : args) {
+      checkSQLiteResult(sqlite3_bind_text(pStmt.get(), 1, arg.getString().c_str(), -1, SQLITE_TRANSIENT));
+      checkSQLiteResult(sqlite3_step(pStmt.get()));
+      checkSQLiteResult(sqlite3_reset(pStmt.get()));
+    }
+  }
+  callback({});
+}
+
+void AsyncStorageModuleWin32::clear(folly::dynamic, Callback callback) {
+  Exec(u8"DELETE FROM AsyncLocalStorage");
+  callback({});
+}
+
+void AsyncStorageModuleWin32::getAllKeys(folly::dynamic, Callback callback) {
+  folly::dynamic jsRetVal = folly::dynamic::array;
+  auto getAllKeysCallback = [&](int cCol, char **rgszColText, char **) {
+    if (cCol >= 1) {
+      jsRetVal.push_back(rgszColText[0]);
+    }
+    return SQLITE_OK;
+  };
+
+  Exec(u8"SELECT key FROM AsyncLocalStorage", getAllKeysCallback);
+
+  if (jsRetVal.size() == 0) {
+    callback({"AsyncStorageError - No Keys Found", {}});
+  }
+}
+
+auto AsyncStorageModuleWin32::PrepareStatement(const char *stmt) -> Statement {
+  sqlite3_stmt *pStmt;
+  checkSQLiteResult(sqlite3_prepare_v2(m_db, stmt, -1, &pStmt, nullptr));
+  return {pStmt, &sqlite3_finalize};
+}
+
+void AsyncStorageModuleWin32::Exec(const char *statement, ExecCallback callback, void *pv) {
+  ExecImpl(m_db, statement, callback, pv);
+}
+
+void AsyncStorageModuleWin32::BindString(const Statement &stmt, int index, const std::string &str) {
+  checkSQLiteResult(sqlite3_bind_text(stmt.get(), index, str.c_str(), -1, SQLITE_TRANSIENT));
+}
+
+void AsyncStorageModuleWin32::checkSQLiteResult(int sqliteResult) {
+  switch (sqliteResult) {
+    case SQLITE_OK:
+    case SQLITE_ROW:
+    case SQLITE_DONE:
+      break;
+    default: {
+      char errorCode[16 + 10 + 1];
```

Comment:
If we're going to put this into a heap allocated buffer later anyway, could we do this the safe way and just use folly::sformat?

Reply:
What's unsafe about how its written now? But sure I can format directly into the heap allocated storage.

Reply:
I guess not unsafe, since sprintf_s will template to buffer size, just annoying since you would need to manually change hardcoded buffer length if changing the string.